Test/94

SungchoonPark opened the pull request:
## #️⃣  연관된 이슈

> #94 
## 📝  작업 내용

> **기업 간단 정보 조회** API 에 대한 Controller, Service, Repository 단 테스트를 진행했습니다.

## 💬  리뷰 요구사항

> 조회 API 이다 보니 Service Layer의 코드가 상대적으로 간단합니다. (단순 repository 코드 호출)
> 따라서 다음과 같은 사안만 테스트 했는데 이게 유효한 테스트일지 궁금합니다.
> 
> 1️⃣ repository 의 테스트가 1번 호출되었는가? `times`이용
> 2️⃣ repository 에서 리턴된 DTO가 expected와 actual이 같은지

SungchoonPark committed:
test(#94): 기업 상세 정보조회에 대한 Service Layer 테스트 코드 작성 완료

SungchoonPark committed:
test(#94): 기업 상세 정보조회에 대한 Repository Layer 테스트 코드 작성 완료

Detail
- 간단 정보 조회 성공에 대한 테스트
- 존재하지 않는 기업에 대한 예외 테스트

SungchoonPark committed:
test(#94): 기업 상세 정보조회에 대한 Controller Layer 테스트 코드 작성 완료

Detail
- 간단 정보 조회 성공에 대한 테스트
- 존재하지 않는 기업에 대한 예외 테스트

2sg reviewed (commented):
지금 코멘트를 기반으로 전체적으로 셀프리뷰 해봐도 좋겠습니다.

SungchoonPark commented on src/test/java/org/choon/careerbee/domain/company/repository/custom/CompanyCustomRepositoryImplTest.java:
```diff
@@ -117,4 +172,13 @@ private Company createCompany(
             .build();
     }
+
+    private Member createMember() {
```
해당 리뷰를 참고하여 fixture 패키지를 생성 후 CompanyFixture에 createCompany를 정적 메서드로 두고 사용하는 방식으로 변경하였습니다.

SungchoonPark committed:
fix(#94): fixture 패키지 생성 및 createCompany 정적 메서드를 이용하여 코드 재활용성을 높임

SungchoonPark committed:
refactor(#94): 테스트의 외부 의존성 제거하여 독립적인 테스트 구성

Before
- DB의 company 데이터값이 달라지면 테스트코드에 영향이 감

After
- 테스트코드 내에서 테스트 데이터셋을 미리 구성해놓는 방식으로 독립적인 테스트 구성
- 반경 외, 내의 데이터셋을 준비하여 실제로 쿼리가 의도대로 동작하는지 테스트

SungchoonPark commented on src/test/java/org/choon/careerbee/domain/company/controller/CompanyControllerTest.java:
```diff
@@ -37,6 +37,37 @@ class CompanyControllerTest {
+    @Test
+    @DisplayName("유효하지 않은 id인 경우 404 에러 발생")
+    void fetchCompanySummary_shouldReturn404_whenCompanyNotFound() throws Exception {
+        Long invalidCompanyId = 1000L;
```
네. 해당 리뷰 반영하여
**유효하지 않은 -> 존재하지 않는** 으로 변경하였으며 테스트 데이터 또한 직접 given 절에서 주입해주어 독립적인 테스트로 변경하였습니다.

SungchoonPark committed:
Merge branch 'develop' of https://github.com/100-hours-a-week/3-team-CareerBee-be into test/94

SungchoonPark committed:
fix(#94): 유효하지 않은 데이터에 대한 테스트를 명확히 구분하도록 변경

Detail
- String 또는 공백 입력을 "유효하지 않은 입력"으로 구분
- 존재하지 않는 기업의 id를 "존재하지 않는 기업"으로 구분

SungchoonPark committed:
add(#94): Controller 단에서 발생하는 PathVariable 관련 예외 처리 핸들러 추가

Detail
- PathVariable을 누락하거나 convert 되지 않는 타입으로 입력됐을 때 예외 처리 핸들러 추

SungchoonPark committed:
chore(#94): test 환경변수 추가

SungchoonPark commented on src/test/java/org/choon/careerbee/domain/company/service/CompanyQueryServiceImplTest.java:
```diff
@@ -59,6 +61,31 @@ void fetchCompaniesByDistance_shouldCallRepository_andReturnExpectedResult() {
+            1L,
+            List.of()
+        );
+        when(companyQueryService.fetchCompanySummary(anyLong())).thenReturn(expectedResponse);
```
service 코드가 아닌 **repository 코드를 모킹**하는 것으로 변경하여 온전히 service 내부 로직은 실행되도록 변경하였습니다.

SungchoonPark committed:
chore(#94): repository 코드를 모킹하여 온전히 Service Layer의 내부 로직을 테스트할 수 있도록 함

SungchoonPark commented on src/test/java/org/choon/careerbee/domain/company/service/CompanyQueryServiceImplTest.java:
```diff
@@ -59,6 +61,31 @@ void fetchCompaniesByDistance_shouldCallRepository_andReturnExpectedResult() {
+        CompanySummaryInfo actualResponse = companyQueryService.fetchCompanySummary(companyId);
+
+        // then
+        verify(companyRepository, times(1)).fetchCompanySummaryInfoById(companyId);
```
리뷰 반영하여 ArgumentCaptor 를 이용한 실제 모킹 객체에 어떤 파라미터가 사용되었는지 검증하였습니다.

SungchoonPark committed:
test(#94): ArgumentCaptor를 활용해 모킹 객체에 어떤 파라미터가 사용됐는지 검증

SungchoonPark commented on src/test/java/org/choon/careerbee/domain/company/service/CompanyQueryServiceImplTest.java:
```diff
@@ -59,6 +61,31 @@ void fetchCompaniesByDistance_shouldCallRepository_andReturnExpectedResult() {
+
     @Test
     @DisplayName("기업 상세 조회 시 repository 호출 및 결과 반환")
     void fetchCompanyDetail_ShouldReturnDetailResponse() {
```
response의 동등성 뿐 아니라 응답 객체의 개별 필드도 검증하는 코드를 추가하였습니다.

SungchoonPark commented on src/test/java/org/choon/careerbee/domain/company/service/CompanyQueryServiceImplTest.java:
실수였습니다. assertJ를 통일화해서 사용하도록 하겠습니다.

SungchoonPark committed:
add(#94): fixture 객체를 통한 코드 재사용성 향상 및 테스트 코드 가독성 향

SungchoonPark committed:
chore(#94): 파일끝에 개행문자를 삽입하여 POSIX 표준 준

SungchoonPark commented on src/test/java/org/choon/careerbee/domain/company/service/CompanyQueryServiceImplTest.java:
```diff
@@ -59,6 +61,31 @@ void fetchCompaniesByDistance_shouldCallRepository_andReturnExpectedResult() {
         assertThat(actualResponse).isEqualTo(expectedResponse);
```
@2sg 현재 테스트 대상이 되는 서비스 코드는 단순히 Repository를 호출하고, 해당 결과를 그대로 반환하는 역할만 수행하고 있습니다.

따라서 테스트 코드에서도 when().thenReturn()을 통해 Mock된 결과를 지정하고, 이를 그대로 검증하게 되는데, 이 방식이 과연 유의미한 테스트인지에 대해 의문이 듭니다.

이런 구조에서는 `ArgumentCaptor`를 활용한 **파라미터 값 검증**과 `verify(..., times(...))`를 통한 **호출 여부 및 횟수 검증** 정도가 더 적절한 테스트 방향이라는 생각이듭니다. 이러한 생각이 바른 방향인지 의문이 생깁니다!

SungchoonPark committed:
test(#94): 모킹한 응답 데이터의 검증로직 추가

SungchoonPark committed:
test(#94): 중복되는 verify 제
